File: elixir/space-age/lib/space_age.ex
defmodule SpaceAge do
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @earth_years 60 * 60 * 24 * 365.25

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(planet, seconds) do
    case planet do
      :earth -> {:ok, earth_age(seconds)}
      :mercury -> {:ok, earth_age(seconds) / 0.2408467}
      :venus -> {:ok, earth_age(seconds) / 0.61519726}
      :mars -> {:ok, earth_age(seconds) / 1.8808158}
      :jupiter -> {:ok, earth_age(seconds) / 11.862615}
      :saturn -> {:ok, earth_age(seconds) / 29.447498}
      :uranus -> {:ok, earth_age(seconds) / 84.016846}
      :neptune -> {:ok, earth_age(seconds) / 164.79132}
      _ -> {:error, "not a planet"}
    end
  end

  defp earth_age(seconds), do: seconds / @earth_years
end

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge
  ( Planet (..),
    ageOn,
  )
where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

earthAge :: Float -> Float
earthAge seconds = seconds / 31557600

ageOn :: Planet -> Float -> Float
ageOn Earth seconds = earthAge seconds
ageOn Mercury seconds = earthAge seconds / 0.2408467
ageOn Venus seconds = earthAge seconds / 0.61519726
ageOn Mars seconds = earthAge seconds / 1.8808158
ageOn Jupiter seconds = earthAge seconds / 11.862615
ageOn Saturn seconds = earthAge seconds / 29.447498
ageOn Uranus seconds = earthAge seconds / 84.016846
ageOn Neptune seconds = earthAge seconds / 164.79132
